src/resolvelib/providers.py

```diff
@@ -136,3 +136,36 @@ def get_dependencies(self, candidate: CT) -> Iterable[RT]:
         specifies as its dependencies.
         """
         raise NotImplementedError
+
+    def narrow_backtrack_selection(
+        self,
+        unsatisfied_names: Iterable[KT],
+        resolutions: Mapping[KT, CT],
+        candidates: Mapping[KT, Iterator[CT]],
+        information: Mapping[KT, Iterator[RequirementInformation[RT, CT]]],
+        backtrack_causes: Sequence[RequirementInformation[RT, CT]],
+    ) -> Iterable[KT]:
+        """
+        Narrows the selection of unsatisfied dependency names during the
+        backtracking process.
+
+        The return value must be a non-empty subset of `unsatisfied_names`,
+        with the simplest implementation being to return `unsatisfied_names`
+        unchanged.
+
+        This method can be used by the provider to optimize the dependency
+        resolution process. `get_preference` will only be called on the
+        names returned from this method. If there is only one name returned
+        from this method, `get_preference` won't be called at all.
+
+        Serving a similar purpose as `get_preference`, this method allows
+        the provider to guide resolvelib through the backtracking process.
+        This method should be used instead of `get_preference` when the
+        provider needs to consider multiple unsatisfied dependency names
+        simultaneously, or when the provider wants to skip checking all
+        names, e.g., because the checks are prohibitively expensive.
+
+        Returns:
+            Iterable[KT]: A non-empty subset of `unsatisfied_names`.
```

Comment:
Could the docstring be extended to describe the meaning of the various arguments? For example, "information" doesn't really describe what the provider can expect to find in that value... 😉

Reply:
Oh yeah, since this method now matches `get_preference` almost entirely (it didn't when I opened this PR) I've updated the docstring to far more closely match `get_preference`, including pulling in the parameter descriptions from there.

I also updated the method name tp `narrow_requirement_selection`, as it fits the docstring descriptions better now, and I like that `backtrack` isn't part of the name (in fact I would like to move backtrack from the parameter name as well, but that would be a breaking change for `get_preference`, I was the one that added that parameter name but I wasn't thinking about resolvelib in a wider context when I added it).